libdnf/dnf-context.cpp

```diff
@@ -1566,8 +1566,13 @@ void
 dnf_context_set_enable_filelists (DnfContext     *context,
                                   gboolean        enable_filelists)
 {
-    DnfContextPrivate *priv = GET_PRIVATE(context);
-    priv->enable_filelists = enable_filelists;
+    // TODO(jkolarik): Simplified as we only accept filelists for optional_metadata_types
+    auto & optional_metadata_types_option = libdnf::getGlobalMainConfig(false).optional_metadata_types();
+    if (enable_filelists) {
+        optional_metadata_types_option.set(libdnf::Option::Priority::RUNTIME, std::vector<std::string>{"filelists"});
+    } else {
+        optional_metadata_types_option.set(libdnf::Option::Priority::RUNTIME, std::vector<std::string>());
+    }
```

Comment:
I have a bad news. If I am not mistaken this approach affects all context instances. I create a unit test that fails with the new implementation.

```
void ContextTest::testMultiContext() {
    auto * context_second = dnf_context_new();
    g_assert(dnf_context_get_enable_filelists(context) == TRUE);
    g_assert(dnf_context_get_enable_filelists(context_second) == TRUE);
    dnf_context_set_enable_filelists(context, FALSE);
    g_assert(dnf_context_get_enable_filelists(context) == FALSE);
    g_assert(dnf_context_get_enable_filelists(context_second) == TRUE);

    g_object_unref(context_second);
}
```

Reply:
My recommendation would be to not modify the global config. But only use it for a default values.

Reply:
The problem is related to previous note - https://github.com/rpm-software-management/libdnf/pull/1635#discussion_r1429735072

Reply:
May I ask you to include the unit test to the PR?

Reply:
OK, I see. I was confused why some other `dnf_context_set_*` and `dnf_context_get_*` methods are directly using the global config object, but now I understand this is a wrong approach. I will rework the logic regarding filelists there and for the other options changing the global config object, I'll create an issue.

Reply:
Follow-up issue: https://github.com/rpm-software-management/libdnf/issues/1640.